base/server/src/main/java/com/netscape/cms/realm/PKIPostgreSQLRealm.java

```diff
@@ -0,0 +1,411 @@
+
+        logger.info("Getting user for cert " + certID);
+
+        String sql = statements.getProperty("getUserByCertID");
+        logger.info("SQL: " + sql);
+
+        try (PreparedStatement ps = connection.prepareStatement(sql)) {
+            ps.setString(1, certID);
+
+            try (ResultSet rs = ps.executeQuery()) {
+
+                if (!rs.next()) {
+                    return null;
+                }
+
+                return createUser(rs);
+            }
+        }
+    }
+
+    public List<X509Certificate> getUserCerts(String userID) throws Exception {
+
+        logger.info("Getting certs for user " + userID);
+
+        String sql = statements.getProperty("getUserCerts");
+        logger.info("SQL: " + sql);
+
+        List<X509Certificate> results = new ArrayList<>();
+
+        try (PreparedStatement ps = connection.prepareStatement(sql)) {
+            ps.setString(1, userID);
+
+            try (ResultSet rs = ps.executeQuery()) {
+                while (rs.next()) {
+                    byte[] data = rs.getBytes("data");
+                    results.add(new X509CertImpl(data));
+                }
+            }
+        }
+
+        return results;
+    }
+
+    public List<String> getUserRoles(String userID) throws Exception {
+
+        logger.info("Getting roles for user " + userID);
+
+        String sql = statements.getProperty("getUserRoles");
+        logger.info("SQL: " + sql);
+
+        List<String> roles = new ArrayList<>();
+
+        try (PreparedStatement ps = connection.prepareStatement(sql)) {
+            ps.setString(1, userID);
+
+            try (ResultSet rs = ps.executeQuery()) {
+
+                while (rs.next()) {
+                    String rolename = rs.getString("group_id");
+                    roles.add(rolename);
+                }
+            }
+        }
+
+        return roles;
+    }
+
+    @Override
+    public Principal authenticate(String username, String credentials) {
+
+        logger.info("Authenticating user " + username + " with password");
+
+        try {
+        connect();
+
+        logger.info("Searching for user " + username);
+        User user = getUserByID(username);
+
+        if (user == null) {
+            logger.warn("Unable to authenticate user " + username + ": User not found");
+            return null;
+        }
+
+        logger.info("Validating password for user " + username);
+        String storedCredentials = user.getPassword();
+
+        if (!handler.matches(credentials, storedCredentials)) {
+            logger.warn("Unable to authenticate user " + username + ": Invalid password");
+            return null;
+        }
+
+        logger.info("User " + username + " authenticated");
+
+        List<String> roles = getUserRoles(username);
+        return new PKIPrincipal(user, null, roles);
+        } catch (Exception e) {
+            logger.error("Problem to cannect the DB and verify the user credentials");
```

Comment:
I think the code within this `try-catch` block should be indented. Could you also wrap and rethrow the original exception?